default/AI/ResourcesAI.py

```diff
@@ -211,30 +381,227 @@ def assess_protection_focus(pid):
+                    curTargetPP += CI - 1 - research_penalty
+                    curTargetRP += CR + 1
+                    focus_manager.bake_future_focus(pid, RFocus, False)
+                continue
+            if adj_round == 4: #assume default IFocus
+                curTargetPP += II  # icurTargets initially calculated by Industry focus, which will be our default focus
+                curTargetRP += IR
+                ratios.append((thisFactor, pid, pinfo))
+
+    ratios.sort()
+    printedHeader = False
+    gotAlgo = tech_is_complete("LRN_ALGO_ELEGANCE")
+    for ratio, pid, pinfo in ratios:
+        do_research = False  # (focus_manager.new_foci[pid]==RFocus)
+        if (priorityRatio < (curTargetRP / (curTargetPP + 0.0001))) and not do_research:  # we have enough RP
+            if ratio < 1.1 and foAI.foAIstate.aggression > fo.aggression.cautious:  # but wait, RP is still super cheap relative to PP, maybe will take more RP
+                if priorityRatio < 1.5 * (curTargetRP / (curTargetPP + 0.0001)):  # yeah, really a glut of RP, stop taking RP
+                    break
+            else:  # RP not super cheap & we have enough, stop taking it
+                break
+        II, IR = pinfo.possible_output[IFocus]
+        RI, RR = pinfo.possible_output[RFocus]
+        # if focus_manager.current_focus[pid] == MFocus:
+        # II = max( II, focus_manager.possible_output[MFocus][0] )
+        if (not do_research and (
+                (ratio > 2.0 and curTargetPP < 15 and gotAlgo) or
+                (ratio > 2.5 and curTargetPP < 25 and II > 5 and gotAlgo) or
+                (ratio > 3.0 and curTargetPP < 40 and II > 5 and gotAlgo) or
+                (ratio > 4.0 and curTargetPP < 100 and II > 10) or
+                ((curTargetRP + RR - IR) / max(0.001, curTargetPP - II + RI) > 2 * priorityRatio))):  # we already have algo elegance and more RP would be too expensive, or overkill
+            if not printedHeader:
+                printedHeader = True
+                print "Rejecting further Research Focus choices as too expensive:"
+                print "%34s|%20s|%15s |%15s|%15s |%15s |%15s" % ("                      Planet ", " current RP/PP ", " current target RP/PP ", "current Focus ", "  rejectedFocus ", " rejected target RP/PP ", "rejected RP-PP EQF")
+            oldFocus = pinfo.current_focus
+            cPP, cRP = pinfo.current_output[IFocus], pinfo.current_output[RFocus]
+            otPP, otRP = pinfo.possible_output[oldFocus]
+            ntPP, ntRP = pinfo.possible_output[RFocus]
+            print "pID (%3d) %22s | c: %5.1f / %5.1f | cT: %5.1f / %5.1f |  cF: %8s | nF: %8s | cT: %5.1f / %5.1f | %.2f" % (pid, pinfo.planet.name, cRP, cPP, otRP, otPP, fociMap.get(oldFocus, 'unknown'), fociMap[RFocus], ntRP, ntPP, ratio)
+            continue  # RP is getting too expensive, but might be willing to still allocate from a planet with less PP to lose
```

Comment:
Commit [0e8bd8b676f45b 'Fix overly indented code in set_planet_industry_and_foci' ](https://github.com/freeorion/freeorion/pull/580/commits/0e8bd8b676f45b05d084149d841880f1bf5f4d01) fixes the bug.

Previously, the code to prevent planets satisfying the conditions now at line 568-573 was indented to the depth inside the `if` of the line  574 `if not printedHeader`.  It was only executed when the header was printed, instead of for each planet that satisfied the entry conditions.

The important (non reporting) line is the `continue` on line 583, which needs to run to skip the next planet.  In the buggy code it was falling through and having its focus set th RFocus.